### DISCUSSION TITLE
Best practices for schemas and organization in django-ninja

### DISCUSSION THREAD

**Discussion by metersk** [vitalik/django-ninja#655] [Q&A]
I'm quite new to REST development and `django-ninja`, but I am enjoying it so far.  My application is becoming more complex and I have some questions on how to organize schemas.  Let's say I have the following models:

```
class Project(TimeStampMixin, models.Model):
    project_name = models.TextField(unique=True)
    project_short_code = models.TextField(unique=True)
    status = models.CharField(max_length=100)
    project_started = models.DateTimeField()

class Experiment(TimeStampMixin, models.Model):
    experiment_name = models.TextField(unique=True)
    experiment_type = models.TextField()
    project = models.ForeignKey(
        Project, on_delete=models.CASCADE, related_name="experiments"
    )

class Plate(TimeStampMixin, models.Model):
    experiment = models.ForeignKey(
        Experiment, on_delete=models.CASCADE, related_name="plates"
    )
    plate_name = models.TextField(unique=True, null=False)
    size = models.IntegerField(null=False)

class Well(TimeStampMixin, models.Model):
    plate = models.ForeignKey(Plate, on_delete=models.CASCADE, related_name="wells")
    row = models.TextField(null=False)
    column = models.TextField(null=False)

class StrainTube(TimeStampMixin, models.Model):
    name = models.TextField(null=False)
    strain_id = models.TextField(null=False)
    is_latest = models.BooleanField(null=False)

class WellContent(TimeStampMixin, models.Model):
    well = models.ForeignKey(
        Well, on_delete=models.CASCADE, related_name="well_contents"
    )
    strain_tube = models.ForeignKey(
        StrainTube, on_delete=models.RESTRICT, related_name="strain_tubes"
    )
```

Summarized, A `project` can have many `experiments`, and `experiment` can have many `plates`,  a `plate` can have many `wells` and a `well` can have many `well_contents`.

As a bit of a preface, I had read somewhere that it is bad practice to use `Optional` when defining schemas, so I tried not to do this wherever possible.  Maybe this is not a totally correct approach though.

The schemas start to get messy when I create views to return different formulations of my data.  Let's say I have a endpoint that returns a plate and it's contents.  The schemas might look like this:

```
class StrainTubeSchemaOut(Schema):
    id: str
    strain_id: str
    is_latest: bool


class WellContentsSchemaOut(Schema):
    well_id: str
    strain_tube: StrainTubeSchema = Field(..., alias="strain_tube")


class WellSchemaOut(Schema):
    id: str
    row: str
    column: str
    well_contents: Optional[List[WellContentsSchema]] = Field(
        ..., alias="well_contents"
    )

class PlateSchemaOut(Schema):
    id: str
    plate_name: str
    size: str
    wells: List[WellSchema] = Field(..., alias="wells")
```

This works fine, but this returns quite a lot of data, so I don't always want to return it.  Let's say now I want to return all `plate_id`s for a given `experiment`.  I might have something like this, where I have a truncated version of my original `plates` schema:

```
class ExperimentPlatesSchema(Schema):
    id: str
    plate_name: str
    size: str

class ExperimentSchema(Schema):
    id: str
    experiment_name: str
    plates: List[PlateExperimentSchemaOut] = Field(..., alias="plates")
 ```

And now let's say I need to create plates, so I have another schema that maps to a plate input, without the `id`:
```
class PlatesSchemaIn(Schema):
    plate_name: str
    size: str
```

So now I'm left with 3 schemas that are related to `plates` and this is only the beginning of fetching and creating data.  It already seems quite messy to me.  I'm finding naming hard as well.  

One solution I was thinking of is using `Optional` fields and fully describing each model with a schema and then truncating the data returned from models, but I'm not sure if that is a best practice.

Thanks for reading, really just looking for any advice on schema organization/naming conventions for a more complex REST API.

**Comment by Jragon** [comment]
> As a bit of a preface, I had read somewhere that it is bad practice to use Optional when defining schemas, so I tried not to do this wherever possible. Maybe this is not a totally correct approach though.

I think this is just because you can use `str | None` with py3.10+ which is a bit less misleading. Since Optional should really mean Nullable (noneable?)


Personally I'm a fan of verbose schemas and recreating them often for different endpoints. Especially as your api grows in complexity it becomes difficult to have one schema for plates and one for experiments etc. We're moving away from this approach and django rest framework. It's a real breath of fresh air since everything is just much more clearly defined. 

In terms of naming, if you need a specific schema for a specific endpoint like `create_plates` I'd have a schema called `CreatePlateIn(Schema)`. 

I agree though it does get unwildy quite quickly. But that is the downside to working within python :shrug: it would be great to be able to define anon schemas like when defining typescrip types
